crates/webidl/src/util.rs

```diff
@@ -841,69 +1073,111 @@ impl<'src> FirstPassRecord<'src> {
-
-        let rust_name = rust_ident(
-            &if overloaded && !arguments.is_empty() {
-                let mut argument_type_names = String::new();
-                for arg in arguments {
-                    let arg = match arg {
-                        Argument::Single(single) => single,
-                        Argument::Variadic(_) => return None,
-                    };
-                    if argument_type_names.len() > 0 {
-                        argument_type_names.push_str("_and_");
-                    }
-                    if same_argument_names {
-                        arg.type_.type_.type_to_string(self, &mut argument_type_names);
-                    } else {
-                        argument_type_names.push_str(&arg.identifier.0.to_snake_case());
-                    }
+        let rust_name = if overloaded && !arguments.is_empty() {
+            let mut argument_type_names = String::new();
+            for argument in arguments {
+                let argument = match argument {
+                    Argument::Single(single) => single,
+                    Argument::Variadic(_) => return None,
+                };
+                if argument_type_names.len() > 0 {
+                    argument_type_names.push_str("_and_");
                 }
-                if name == "new" {
-                    "with_".to_owned() + &argument_type_names
+                if same_argument_names {
+                    argument.type_.type_.push_type_name(self, &mut argument_type_names);
                 } else {
-                    name.to_snake_case() + "_with_" + &argument_type_names
+                    argument_type_names.push_str(&argument.identifier.0.to_snake_case());
                 }
+            }
+            if name == "new" {
+                "with_".to_owned() + &argument_type_names
             } else {
-                name.to_snake_case()
+                name.to_snake_case() + "_with_" + &argument_type_names
             }
-        );
-        let name = name.to_string();
+        } else {
+            name.to_snake_case()
+        };
 
         let js_ret = ret.clone();
 
         if catch {
             ret = Some(ret.map_or_else(|| result_ty(unit_ty()), result_ty))
         }
 
-        let shim = {
-            let ns = match kind {
-                backend::ast::ImportFunctionKind::Normal => "",
-                backend::ast::ImportFunctionKind::Method { ref class, .. } => class,
+        let variants = self.get_variants(arguments, &kind)?;
+        let multiple_variants = variants.len() > 1;
+        let mut arguments_count_variants_count = BTreeMap::new();
+        for variant in &variants {
+            arguments_count_variants_count
+                .entry(variant.len())
+                .and_modify(|variants_count| { *variants_count += 1usize; })
+                .or_insert(1usize);
+        }
+        let mut result = Vec::new();
+        for variant in variants {
```

Comment:
This function is getting pretty huge -- could we factor out some bits to make it less monolithic and more readable?